0004-predicate-types.md

```diff
@@ -0,0 +1,192 @@
+  var j: Odd := if Odd(i) then i else 1;
+  j := j + 1; // '+' highlighted, value does not satisfy the subset constraints of 'Odd'
+  j := j + 2; // Here you get no error, assuming j is odd.
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Ideally, we would wait for `predicate` to be compilable by default, so that the keyword `type` is unambiguously added afterwards: `ghost predicate type` or `predicate type`
+(contrary to `predicate type` and `predicate method type` that would be weird today)
+
+We would have to change:
+* **Dafny AST:** Add fields `isType` and `witness` to `Predicate`.
+* **Parser:** support the new syntax with the optional fields above, namely an optional "type" after `predicate` and an optional `witness ...` like subset types after the definition of the predicate.
+  We should reject `predicate type`s with more than one argument (for now), with `requires` or `reads`, but we can keep `ensures`
+* **Resolver:** Add an internal rewriter before resolution that converts a `predicate type` into both a `predicate`,
+  and a subset type of the same name with a call to the predicate for constraint, with the witness provided.
+  We also need a predicate to be able to have the same name as a type, for example by internally prefixing types with "t_"
+
+There is nothing more to implement for the verifier and compiler, since this is desugarding.
+The last example above would be rewritten:
+
+```dafny
+predicate Odd(x: int) {
+  x % 2 == 1
+}
+type Odd = x: int | Odd(x) witness 1
+
+method Main() {
+  var i:int := f();
+  var j: Odd := if Odd(i) then i else 1;
+  j := j + 1; // You'd get an error here, cannot assign something not Odd to something that is Odd
+  j := j + 2; // Here you get no error, assuming j is odd.
+}
+```
+
+which would automatically satisfy all 
+
+## Further notes
+
+* We would still be able to use `{:opaque}` and use `reveal` because that annotation can be directly added after `predicate`.
+* It would be possible to define `predicate newtype` the same way, and it would have the same advantages.
+* In a second milestone, we could define another kind of `predicate type` inside a class or a datatype,
+  where the type is statically accessible, but predicate is not and has no argument. For example:
+  ```dafny
+  datatype Boxed = BoxedNumber(n: int) {
+    predicate type Odd() {
+      this.n % 2 == 1
+    }
+  }
+  ... var j: Boxed.Odd = BoxedNumber(1) ...
+  ```
+  would become
+  ```dafny
+  datatype Boxed = BoxedNumber(n: int) {
+    predicate Odd() {
+      this.n % 2 == 1
+    }
+    type Odd = s: Boxed | s.Odd()
+  }
+  ... var j: Boxed.Odd = BoxedNumber(1) ...
+  ```
+  Dafny does not support nested types (which would automatically be static) so that would be a huge refactoring.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* It feels odd to have restrictions on a predicate so that it becomes a type.
+* It might be even more confusing for newcomers.
```

Comment:
To teach beginners, I'm guessing it will be preferable to first teach about subset types and about predicates. When those two concepts are clearly understood, one can introduce `predicate type`, which then acts just like a shorthand.